The initial steps remain the same, where we import the necessary libraries, separate the dependent and independent variables, and split the dataset.

In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import ElasticNetCV

In [2]:
os.chdir('C:/Users/Admin/Downloads')

In [3]:
# read the data
df = pd.read_csv("Energy_Efficiency_Overfit_Dataset_Updated.csv")

In [4]:
# separate the dependent and independent variables
X = df.drop('Energy_Efficiency_Rating', axis = 1)
y = df['Energy_Efficiency_Rating']

In [5]:
# split the dataset into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [6]:
# create an array with 20 numbers equllay spaced between 0 to 10
alphas = np.linspace(0, 10, 20)
alphas

array([ 0.        ,  0.52631579,  1.05263158,  1.57894737,  2.10526316,
        2.63157895,  3.15789474,  3.68421053,  4.21052632,  4.73684211,
        5.26315789,  5.78947368,  6.31578947,  6.84210526,  7.36842105,
        7.89473684,  8.42105263,  8.94736842,  9.47368421, 10.        ])

One thing to note while using cross-validation with an elastic net is the L1 ratio parameter. The L1 ratio parameter controls the mix between L1, which is lasso, and L2, which is ridge regularization. In our code, we have used a list of ratios to consider. Example, 0.1, 0.5, 0.7, etc. An L1 ratio of one corresponds to pure L1 regularization, while an L1 ratio of zero corresponds to pure L2 regularization. By specifying a range of values, you allow the model to select the best mix of L1 and L2 regularization through cross-validation. Let's run the code below and fit it on the training data.

In [7]:
# Initialize ElasticNetCV to find the best alpha and l1_ratio
elastic_net_cv = ElasticNetCV(alphas=alphas, cv=10, max_iter=10000, l1_ratio=[.1, .5, .7, .9, .95, .99, 1])
elastic_net_cv.fit(X_train, y_train)

C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1195.924420487514, tolerance: 4.484421350867123
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Ob

C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1227.6188012321218, tolerance: 4.40692803260468
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1118.1932662646723, tolerance: 4.849214215897236
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning

C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1262.7197417464413, tolerance: 4.523553771776862
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: O

C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1142.8477133891574, tolerance: 4.846841737561287
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1177.5033624545395, tolerance: 4.639576483655909
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarnin

C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1262.719741745932, tolerance: 4.523553771776862
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning: Coordinate descent without L1 regularization may lead to unexpected results and is discouraged. Set l1_ratio > 0 to add L1 regularization.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 1290.060019784003, tolerance: 4.822783202358667
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Admin\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:614: UserWarning:

ElasticNetCV(alphas=array([ 0.        ,  0.52631579,  1.05263158,  1.57894737,  2.10526316,
        2.63157895,  3.15789474,  3.68421053,  4.21052632,  4.73684211,
        5.26315789,  5.78947368,  6.31578947,  6.84210526,  7.36842105,
        7.89473684,  8.42105263,  8.94736842,  9.47368421, 10.        ]),
             cv=10, l1_ratio=[0.1, 0.5, 0.7, 0.9, 0.95, 0.99, 1],
             max_iter=10000)

In [8]:
# retrieve the best alpha and l1_ratio
best_alpha_enet = elastic_net_cv.alpha_
best_l1_ratio = elastic_net_cv.l1_ratio_

In [9]:
# print the l1_ratio and best_alpha value
(best_l1_ratio, best_alpha_enet)

(0.1, 0.5263157894736842)

In [10]:
# R-squared scores for the Ridge model
r2_train_enet = elastic_net_cv.score(X_train, y_train)
r2_test_enet = elastic_net_cv.score(X_test, y_test)

r2_train_enet, r2_test_enet

(0.9412833484776975, 0.8727794722404819)

With this, we have got the R^2 value for both the training and testing sets. The performance is very similar to what we achieved with the L2 regularization. As a result, comparatively, we can see L1 regularization is the most effective technique in our case because of the smallest difference between the training and testing sets.

In [11]:
# fit the coefficients after elastic net regularization
coefficients_after_enet = pd.Series(elastic_net_cv.coef_, index=X_train.columns)
coefficients_after_enet

Wall_Area              0.299674
Roof_Area              0.245787
Window_Area            0.260984
Overall_Height        -0.001212
Outdoor_Temperature   -0.091724
Humidity              -0.061354
Noise_Feature_1       -0.123267
Noise_Feature_2        0.019370
Noise_Feature_3       -0.075433
Noise_Feature_4        0.125379
Noise_Feature_5        0.099198
Noise_Feature_6        0.171870
Noise_Feature_7       -0.000000
Noise_Feature_8        0.533941
Noise_Feature_9        0.000095
Noise_Feature_10       0.000000
Orientation_East       0.233718
Orientation_North     -0.247003
Orientation_South      0.439144
Orientation_West      -0.425870
Glazing_Type_Type_A    0.000000
Glazing_Type_Type_B    0.000000
Glazing_Type_Type_C   -0.077079
dtype: float64

Lastly, let's check the coefficient of the elastic net model. As we can see, the elastic net has combined the specialties of L1 and L2 regularization, it has reduced the coefficients of some features, whereas completely reducing some coefficients to absolute zero. This is how elastic net is a hybrid solution. 

Comparison of L1, L2, and ElasticNet regularization: Based on the performance, we observe that L1, L2, and elastic net all achieve an R^2 of 0.87 on the test dataset. This is slightly better than the 0.86 achieved by unregularized linear regression. Next we have model complexity. Here L1 regularization produced a model with fewer features. This might be beneficial if model interpretability and simplicity are important. Lastly, let's look at overfitting. The unregularized model showed a difference of 0.9 between train and test scores. L1 showed a difference of 0.05. Both L2 and elastic net showed a difference of 0.067. These figures suggest that L1 regularization is more resilient to overfitting in the case of apex dataset. This suggests that L1 generalizes marginally better to new data compared to L2 and elastic net regularization in this case